### ISSUE TITLE
How about WSL distros as dependencies for APPX. Enabling Store distribution of Linux apps.

### ISSUE THREAD

**Issue by PhMajerus** [Microsoft/WSL#2600]
Now that we have WSL distros in the store as standalone packages, how about making them available as dependencies for other APPX ?
This would enable distributing apps that target Linux APIs in the Store.
Even better, developers could then mix UWP, Win32 and Linux within a single package, providing a UWP frontend that relies on Linux binaries underneath to get some work done for example.

Just like Centennial apps can add a dependency to VC++ redists, Linux apps could be APPX containing only the app's files, a dependency on a shared distro, and a list of distro-specific packages dependencies.
When installing such an app, the user wouldn't see the distro itself, only the installed apps. General-purpose distros for advanced users would then be just launcher APPX that contain the dependency to the distro, common distro-specifc dependencies, and a shortcut to their shell.

This would make it possible to bring Linux apps to the Store by packaging them into an APPX, and work as first-class citizens along UWP and Win32 Centennial apps.
Later, a shared X11 server could be provided to enable GUI apps as well, again as a dependency for Store apps.

And while we're at it...
I know Astoria (basically Android on WSL) has been dropped, but with ARM64 tablets coming soon, it should be reconsidered.
I don't think Astoria should be released on its own as a distro in the Store, but in the same concept as above, it would make a lot of sense to release Astoria as an APPX dependency.
This could enable new apps to be published in the Store using a shared Astoria WSL instance, helping to fill the tablet-optimized app-gap for the upcoming ARM64 tablets with small screens, on which Win32 apps will work, but not be comfortable.
An x86 build of Astoria could also have benefits for Intel-based small computers, as developers would have to recompile their Android app to target it, but wouldn't have to rewrite them to target a completely different API.

**Comment by zadjii-msft** [comment]
*cough* @bitcrazed

**Comment by bitcrazed** [comment]
/cc @tara-raj 

Thanks for your thoughts @PhMajerus & for your enthusiasm :)

A few things here: 

1. Please post asks for futures/ideas on [our UserVoice](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo) which has way better voting features than GitHub, making it easier for us to assess actual demand in the community of proposed ideas
1. Please post one ask/idea at a time. It's really tough to have sensible conversations about several ideas in one post, in a forum like GitHub/UserVoice ;)
1. We are aware of the ask & scenarios for WSL on x86 & ARM64

To your main ask:
1. Certainly not against your suggestion, but there are a number of challenges that'd need to be carefully thought through, including:

    1. A dependency on a Linux distro would be a tough install today for users who don't have WSL enabled because they'd have to allow the installation of WSL itself, which requires a reboot before it can run. We have a backlog item to see if there's a way around this, but it's not currently high on the list vs. all the other things on the backlog
    1. There'd need to be a way for an app publisher to enumerate the distros or distro families that their tool was compatible with - and for the end-user to select and install one of the family of compatible distros
    1. It'd be tough to package Linux apps and tools themselves in APPX's since Linux has no package manager that knows how, for example, to install, uninstall, update, or patch tools installed by an external package installation toolchain.
    1. Etc.

So, in short, great idea. Not against it, but there are significant hurdles getting in the way. Will reconsider in the future.

Closing this issue for now as this isn't really an issue requiring tracking and a fix. Please re-post asks on UserVoice if you'd like :)

**Comment by PhMajerus** [comment]
@bitcrazed 
Ah, didn't think I would get your attention Rich, thanks for taking the time to read and reply.
I am quite enthusiastic indeed, I was hoping SFU would be extended to be Linux-compatible at the source level for over a decade, felt is was the wrong direction to get developers when SUA got dropped, but then you and your team over delivered with the binary compatibility and pico-process architecture.
So congrats to the Lxss and ConhostV2 teams!

If/for when you get back to considering it, I actually think APPX do not need to list distros they support. Developers should set one explicitly and handle dependencies for the one they are compiling for and testing for. Maybe even just package all dependencies not in the distro's standard install within the APPX itself.
Since Lxss can run instances simultaneously, it's fine, and if you let them list several distros, you risk getting into a compatibility nightmare.

Thanks again.

**Comment by bitcrazed** [comment]
TBH, SFU/Interix approached the "run UNIX tools on Windows" in a very odd way, and wouldn't have solved most of the issues we are solving now with WSL. 

Remember - while each Linux ecosystem shares a common/similar kernel and a similar set of core tools, commands, etc., they also differ considerably in package management, configuration, init, etc. Since many tools would need to perform some kind of file-layout / package management / installation, etc., they'll often need to specify which distro family they support.
